File: java_project/oops/NAUGHTEEN/Shoe.java
package oops.NAUGHTEEN;

public class Shoe extends LeatherItems {
    private int sizeOfShoe ;
    private String brand;
    private static int counter =1;

    public Shoe (String itemId, String category, String type, int sizeOfShoe, String brand){
        super (itemId, category, type);
        this.sizeOfShoe = sizeOfShoe;
        this.brand = brand;
    }
    public int getSizeOfShoe(){
        return sizeOfShoe;
    }
    public String getBrand(){
        return brand;
    }
    public boolean validateSizeOfShoe(){
        if (this.sizeOfShoe >= 4 && this.sizeOfShoe <= 12){
            return true;
        }
        return false;
    }
    public boolean validateBrand(){
        if (this.type != null && 
           (this.type.equalsIgnoreCase("Flat") || 
            this.type.equalsIgnoreCase("Heels") || 
            this.type.equalsIgnoreCase("Stilettos") || 
            this.type.equalsIgnoreCase("Wedge"))) {
            return true;
        }
        return false;
    }
    @Override
    public boolean validateType() {
        if (this.type != null && 
           (this.type.equalsIgnoreCase("Flat") || 
            this.type.equalsIgnoreCase("Heels") || 
            this.type.equalsIgnoreCase("Stilettos") || 
            this.type.equalsIgnoreCase("Wedge"))) {
            return true;
        }
        return false; 
    }
    public float identifyDiscountPercentage(){
        if (!validateBrand()){
            System.out.println("Invalid Brand.");
            return 0f;
        }
        float discount = 0f;
        if (this.brand.equalsIgnoreCase("Reebok")) {
            discount = 10.0f;
        } else if (this.brand.equalsIgnoreCase("Nike")) {
            discount = 25.0f;
        } else if (this.brand.equalsIgnoreCase("Adidas")) {
            discount = 40.0f;
        }
        return discount;    
    }
    @Override
    public float identifyCost(){
        if (!validateSizeOfShoe() && validateType()){
            System.out.println("Invalid Shoe Type or Size.");
            return 0f;
        }
        float cost = 0f;
        if (this.sizeOfShoe >=4 && this.sizeOfShoe<=7){
            if (this.type.equalsIgnoreCase("Flat")) cost = 7.0f;
            else if (this.type.equalsIgnoreCase("Heels")) cost = 8.0f;
        } else if (this.sizeOfShoe >= 8 && this.sizeOfShoe <= 12) {
            if (this.type.equalsIgnoreCase("Stilettos")) cost = 9.0f;
            else if (this.type.equalsIgnoreCase("Wedge")) cost = 10.0f;
        }
        
        return cost;
       
    }
     @Override
        public float calculateBillAmount(){
            if (!validateCategory()){
                System.out.println("Invalid Category.");
                return 0f;
            }
            float baseCostDollers = identifyCost();
            if (baseCostDollers <=0){
                System.out.println("Cost cannot be calculated.");
                return 0f;
            }
            float costInRupees = baseCostDollers * 61.04f;
            float discountPrcent = this.identifyDiscountPercentage();
            costInRupees= costInRupees-(costInRupees * discountPrcent / 100);
            float taxPercent = 0f;
            if (this.category.equalsIgnoreCase("Imported")) {
             taxPercent= 20f;
            }
            float finalBill = costInRupees + (costInRupees * taxPercent / 100);
            this.billId = this.brand + this.itemId + counter;
            counter++;

            return finalBill;
        }
}

File: js_project/oops/NAUGHTEEN/Shoe.js
class Shoe extends LeatherItems {
    // Static counter property unique to the Shoe class
    static counter = 1;

    constructor(itemId, category, type, sizeOfShoe, brand) {
        super(itemId, category, type);
        this._sizeOfShoe = sizeOfShoe;
        this._brand = brand;
    }

    get sizeOfShoe() {
        return this._sizeOfShoe;
    }

    get brand() {
        return this._brand;
    }

    validateSizeOfShoe() {
        if (this._sizeOfShoe >= 4 && this._sizeOfShoe <= 12) {
            return true;
        }
        return false;
    }

    validateBrand() {
        if (this.type !== null && 
           (this.type.toLowerCase() === "flat" || 
            this.type.toLowerCase() === "heels" || 
            this.type.toLowerCase() === "stilettos" || 
            this.type.toLowerCase() === "wedge")) {
            return true;
        }
        return false;
    }

    // Overriding validateType abstract method
    validateType() {
        if (this.type !== null && 
           (this.type.toLowerCase() === "flat" || 
            this.type.toLowerCase() === "heels" || 
            this.type.toLowerCase() === "stilettos" || 
            this.type.toLowerCase() === "wedge")) {
            return true;
        }
        return false;
    }

    identifyDiscountPercentage() {
        if (!this.validateBrand()) {
            console.log("Invalid Brand.");
            return 0;
        }
        
        let discount = 0;
        const currentBrand = this._brand ? this._brand.toLowerCase() : "";
        
        if (currentBrand === "reebok") {
            discount = 10.0;
        } else if (currentBrand === "nike") {
            discount = 25.0;
        } else if (currentBrand === "adidas") {
            discount = 40.0;
        }
        return discount;
    }

    identifyCost() {
        if (!this.validateSizeOfShoe() && this.validateType()) {
            console.log("Invalid Shoe Type or Size.");
            return 0;
        }
        
        let cost = 0;
        const currentType = this.type ? this.type.toLowerCase() : "";

        if (this._sizeOfShoe >= 4 && this._sizeOfShoe <= 7) {
            if (currentType === "flat") cost = 7.0;
            else if (currentType === "heels") cost = 8.0;
        } else if (this._sizeOfShoe >= 8 && this._sizeOfShoe <= 12) {
            if (currentType === "stilettos") cost = 9.0;
            else if (currentType === "wedge") cost = 10.0;
        }
        
        return cost;
    }

    calculateBillAmount() {
        if (!this.validateCategory()) {
            console.log("Invalid Category.");
            return 0;
        }
        
        const baseCostDollars = this.identifyCost();
        if (baseCostDollars <= 0) {
            console.log("Cost cannot be calculated.");
            return 0;
        }
        
        let costInRupees = baseCostDollars * 61.04;
        const discountPercent = this.identifyDiscountPercentage();
        costInRupees = costInRupees - (costInRupees * discountPercent / 100);
        
        let taxPercent = 0;
        if (this.category && this.category.toLowerCase() === "imported") {
            taxPercent = 20.0;
        }
        
        const finalBill = costInRupees + (costInRupees * taxPercent / 100);
        
        this._billId = this._brand + this.itemId + Shoe.counter;
        Shoe.counter++;

        return finalBill;
    }
}